### DISCUSSION TITLE
Change homography plane to avoid miniaturization effect

### DISCUSSION THREAD

**Discussion by neriniccolodei** [OpenStitching/stitching#146] [Q&A]
Hello, 

I'm curious if there's a swift method for selecting a reference image to use for warping the others to its plane in the stitching process. In my current project, stitching multiple images of an ArUco board is causing a miniaturization effect, as illustrated in the attached image. Hence, I'm exploring the possibility of minimizing this effect by adjusting the projection plane to match that of the central image within the array of stitched images.

![MiniaturizationEffect](https://github.com/OpenStitching/stitching/assets/62433425/5e7339ea-5081-4531-91fe-f785696bd55c)

Best regards

**Comment by lukasalexanderweber** [comment]
The Images are aligned onto the first image you pass in, so changing that would lead to different results. You can also try different warping techniques

**Reply by neriniccolodei** [reply]
Thank you. I have tried using different warping techniques, but I cannot seem to have anything working as desired. The affine stitcher works a bit better in terms of keeping frames the same size, but as soon as I lower the confidence threshold (e.g., 0.7) I get this type of error (stitching 27 frames, `range_width` set to 1, `final_megapix = 1`, `medium_megapix = 0.3`, `low_megapix = 0.1`).

`cv2.error: OpenCV(4.8.0) D:\a\opencv-python\opencv-python\opencv\modules\core\src\alloc.cpp:73: error: (-4:Insufficient memory) Failed to allocate 2338783274100 bytes in function 'cv::OutOfMemoryError'`

Which should be stating that the amount of data from the stitching method is too big to handle/store.

I imagine trying the "plane" warping would be what I want? 
But the output from that is:
`OpenCL error CL_MEM_OBJECT_ALLOCATION_FAILURE (-4) during call: clEnqueueNDRangeKernel('buildWarpPlaneMaps', dims=2, globalsize=68096x39624x1, localsize=NULL) sync=true
Traceback (most recent call last):
  File "c:\Users\Neri\OneDrive - Scuola Superiore Sant'Anna\EndoProject\VisualStudio\Project\DetailedStitcher.py", line 140, in <module>
    warped_final_imgs = list(warper.warp_images(final_imgs, cameras, camera_aspect))
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Neri\AppData\Local\Programs\Python\Python311\Lib\site-packages\stitching\warper.py", line 41, in warp_images
    yield self.warp_image(img, camera, aspect)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Neri\AppData\Local\Programs\Python\Python311\Lib\site-packages\stitching\warper.py", line 45, in warp_image
    _, warped_image = warper.warp(
                      ^^^^^^^^^^^^
cv2.error: OpenCV(4.8.0) D:\a\opencv-python\opencv-python\opencv\modules\core\src\ocl.cpp:5954: error: (-220:Unknown error code -220) OpenCL error CL_MEM_OBJECT_ALLOCATION_FAILURE (-4) during call: clEnqueueReadBuffer(q, handle=000001DD535EC990, CL_TRUE, 0, sz=10779764712, data=000001E05C608080, 0, 0, 0) in function 'cv::ocl::OpenCLAllocator::map'`

Which should be due to GPU overflow...

**Reply by lukasalexanderweber** [reply]
You can try reducing the img size with final_megapix

**Reply by neriniccolodei** [reply]
Hello, that unfortunately doesn't fix the issue. 

I think the problem is in the `camera_adjuster`. Is it an optimiser? When running without that, nor the wave corrector, the panorama is less precise but at least it computes very fast. 

Another question I'd like to ask is, can I save the homographies between frames somehow? I can't find them in the code. 
I would like to save the relative homographies and compute a "global" homography so that I may manually premultiply images with that so as to center them with respect to my chosen image. 

I'm trying to write a simple stitching file using directly OpenCV functions, but my performances are not as good. I am using brute force matcher, no optimisation, etc.
